### ISSUE TITLE
APSZOS-23 zWLM Classification for EJB Timers

### ISSUE THREAD

**Issue by tkburroughs** [OpenLiberty/open-liberty#9563]
Support context propagation for non-persistent EJB timers.

WAD: https://ibm.box.com/s/simbqb0esurw3rdsfg5j4rbs5j6qstim

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
# Feature Ship Release Readiness Checklist 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## List of Steps to complete or get approvals / sign-offs for Onboarding to the Liberty release (GM date)
 
Instructions:
 - Do the actions below and mark them complete in the checklist when they are done.
 - Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE **Before Development Starts or 8 weeks before Onboarding** 
- [x] POC Design / WAD Review Scheduled (David Chang) or N/A. AHA Feature marked.
- [x] POC Design / WAD Reviewed (Feature Owner) or N/A. AHA Feature marked.
- [x] Complete any To-Do's from the POC Review, and create a new TO-DO in the AHA Feature for the "Design Approval Request" and assign it to Ian Robinson. Leave the date blank. (Feature Owner)
- [x] Design / WAD Approval (Ian Robinson) or N/A. AHA Feature marked.
- [ ] No Design / No WAD Approval (Arthur De Magalhaes - cloud / Alasdair Nottingham - server) or N/A. AHA Feature marked.
- [x] AHA Feature has link to Github epic or issue. Github epic has link back to AHA Feature. (Feature Owner or Backlog Subtribe PM).
- [x] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [x] ID Requirements identified. (Epic owner / Feature owner with ID focal point)
- [x] Create a child task of the epic entitled "FAT Approval Test Summary".  Add and fill in the template as described here: https://github.ibm.com/was-liberty/WS-CD-Open/wiki/Feature-Review-(Feature-Test-Summary-Process)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE **3 weeks before Onboarding** 
- [x] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or **N/A**. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE ** 3 weeks before Onboarding** 
- [x] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE **2 weeks before Onboarding** 
- [x] Implementation complete. (Epic owner / Feature owner) 
- [x] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [x] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## APPROVALS with TARGET COMPLETION DATE **2 to 1 week before Onboarding** 
Prereq: You must have the Design Approval or No Design Approval on the AHA Feature.
- [x] **Accessibility** - (G Scott Johnston). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [x] **FAT Liberty SOE** - (Kevin Smith). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [x] **Globalization** (Marika Joannidis - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.
- [x] **ID** - (Kareen Deen). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [x] **Performance** - (Jared Anderson). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [x] **Serviceability** - (Don Bourne). Serviceability has been addressed. 
- [x] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [x] **SVT** - (Greg Ecock - Cloud, Brian Hanczaryk- APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.
- [x] **Demo** - (Liberty only - Tom Evans or Chuck Bridgham). Demo is scheduled for an upcoming EOI. Approver adds label focalApproved:demo to the Epic in Github.
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## TARGET COMPLETION DATE **1 week before Onboarding** 
- [x] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [ ] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [x] Github Epic and Epic's issues are closed / complete. All PRs are committed to the master branch. (Epic owner / Feature owner / Backlog Subtribe PM)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## NOT REQUIRED FOR A FEATURE
- [x] **OL Guides** - (Yee-Kang Chang). Assessment for OL Guides is complete or **N/A**. 
- [x] **WDT** - (Leonard Theivendra). WDT work complete or **N/A**.
 
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Related Deliverables TARGET COMPLETION DATE **General Availability** 
- [ ] Blog article writeup (Epic owner / Feature owner / Laura Cowen)

**Comment by tkburroughs** [comment]
@gscottj this feature is ready for Accessibility review; this feature is not adding or modifying any user interfaces.

**Comment by marikaj123** [comment]
Does the feature require translation?

**Comment by gscottj** [comment]
This feature don't not update or add any new user interfaces. AVT for CLI interfaces is in good standing. Please be sure that all new documentation to support this feature are also fully tested for compliance.

**Comment by tkburroughs** [comment]
@marikaj123 Yes, translation is required; metatype.properties was merged 2 weeks ago with PR https://github.com/OpenLiberty/open-liberty/pull/8818

File changed is :
https://github.com/OpenLiberty/open-liberty/blob/integration/dev/com.ibm.ws.ejbcontainer.timer/resources/OSGI-INF/l10n/metatype.properties

It does not appear to be back from translation yet; is there a way to confirm it has been sent for translation?

**Comment by skasund** [comment]
@tkburroughs  L2 has requested a live STE presentation for this feature.  The STE template can be found at the link below: 

https://ibm.box.com/s/1an42g7zdgmaj84w7dft0indqfgi8ffm

Please upload the completed slides to the same STE Archive BOX folder, and I will provide the STE approval.  The live presentation can be scheduled with the L2 team at a later date.  Thanks!

**Comment by tkburroughs** [comment]
@skasund The STE has been uploaded to the STE Archive BOX folder.  It is titled `STE-APSZOS-23 zWLM Classification for EJB Timers.ppt` and the box link is : https://ibm.box.com/s/voi3mfopeax1illi9ohjucxowecp8gc2

**Comment by tkburroughs** [comment]
@chirp1 Documentation update has been submitted here : https://github.ibm.com/was-liberty/liberty-docs/issues/1685

**Comment by tkburroughs** [comment]
@donbourne what are the current requirements for `Serviceability` approval?

**Comment by skasund** [comment]
@tkburroughs  Thanks for the STE slides.  I've provided the approval.  I'll have the L2 team reach out to you to set up a live presentation.

**Comment by marikaj123** [comment]
Translation files did not arrive as per schedule. Looking to have them back by Monday.

**Comment by donbourne** [comment]
**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  WAD -- does the WAD identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  
a) What problem paths were tested and demonstrated? 
b) Who did you demo to?
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?  

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.

**Comment by chirp1** [comment]
Approving.

**Comment by mmerali96** [comment]
1. The WAD does not identify the problems that the customer will most likely see.  There is a new configuration for timerService element of the ejbContainer element that is needed to use this feature. The newly added nonPersistentContextServiceRef attribute must point to a contextService element in order propagate a context. If this is set to an invalid context or no context, then no context will be propagated to maintain the default behavior of non persistent EJB Timers.
2. We are giving the demo to the WAS z/OS L2 team on November 19th. The L2 team agreed that the serviceability of problem scenarios is sufficient because it is consistent with serviceability of context propagation on persistent EJB Timers and asynchronous enterprise bean methods. 
3.  Kevin Lefebvre is the SVT focal point for this feature. He stated that an SVT would not be needed for this issue.
4.  WL3EJB will handle PMRs for questions about EJB Containers for this feature. WL3ASB will handle PMRs for concurrency related issues for this feature. LL3ZOS will handle PMRs for issues relating to the zosWLMContext context.

**Comment by marikaj123** [comment]
The returned translated files were picked by the official Release build.

**Comment by tkburroughs** [comment]
@marikaj123 Some of the translated text, specifically the description of a new attribute in metatype was found to violate a 50 character max limit.... so new text must be delivered.  This will of course then need an additional translation.  The epic has been delayed to 20.0.0.1.

**Comment by marikaj123** [comment]
The Globalization approval will be provided after the fix goes into the release build

**Comment by marikaj123** [comment]
According to Mehdi, the beta content were GMed during Liberty beta release 190011.

**Comment by mmerali96** [comment]
Fixed metatype text that failed due to being over the 50 char limit in PR #9963
Translation changes merged to master in #10153

**Comment by mmerali96** [comment]
Removed beta guard in #10023

**Comment by tkburroughs** [comment]
Confirmed removal of beta guard included in 20.0.0.1 build image.  All work is complete... closing.
